os/kernel/binary_manager/binary_manager_data.c

```diff
@@ -206,37 +206,192 @@ binmgr_uinfo_t *binary_manager_get_udata(uint32_t bin_idx)
+			return;
+		}
+	}
+
+#ifdef CONFIG_SUPPORT_COMMON_BINARY
+	if (!strncmp(name, BM_CMNLIB_NAME, sizeof(BM_CMNLIB_NAME))) {
+		bin_idx = BM_CMNLIB_IDX;
+	} else
+#endif
+	{
+		bin_idx = ++g_bin_count;
 	}
 
 	/* Initialize binary table and register binary name, version and priority */
-	g_bin_count++;
-	BIN_ID(g_bin_count) = -1;
-	BIN_FILECNT(g_bin_count) = 1;
-	BIN_RTLIST(g_bin_count) = NULL;
-	BIN_NRTLIST(g_bin_count) = NULL;
-	BIN_STATE(g_bin_count) = BINARY_INACTIVE;
-	BIN_VER(g_bin_count, 0) = version;
-	BIN_LOAD_PRIORITY(g_bin_count, 0) = load_priority;
-	strncpy(BIN_NAME(g_bin_count), name, BIN_NAME_MAX - 1);
-	BIN_NAME(g_bin_count)[BIN_NAME_MAX - 1] = '\0';
-	sq_init(&BIN_CBLIST(g_bin_count));
-
-	bmvdbg("[USER %d] %s\n", g_bin_count, BIN_NAME(g_bin_count));
+	BIN_ID(bin_idx) = -1;
+	BIN_RTLIST(bin_idx) = NULL;
+	BIN_NRTLIST(bin_idx) = NULL;
+	BIN_STATE(bin_idx) = BINARY_INACTIVE;
+	BIN_PARTNUM(bin_idx, 0) = part_num;
+	BIN_PARTSIZE(bin_idx, 0) = part_size;
+	strncpy(BIN_NAME(bin_idx), name, BIN_NAME_MAX - 1);
+	BIN_NAME(bin_idx)[BIN_NAME_MAX - 1] = '\0';
+	sq_init(&BIN_CBLIST(bin_idx));
 
-	return g_bin_count;
+	bmvdbg("[USER%d : 1] %s size %d num %d\n", bin_idx, BIN_NAME(bin_idx), BIN_PARTSIZE(bin_idx, 0), BIN_PARTNUM(bin_idx, 0));
+}
+
+/****************************************************************************
+ * Name: binary_manager_scan_ubin
+ *
+ * Description:
+ *	 This function scans user binaries and update information.
+ *
+ ****************************************************************************/
+int binary_manager_scan_ubin(int bin_idx, bool crc_check)
+{
+	int ret;
+	int part_idx;
+	int latest_idx;
+	int bin_count;
+	uint32_t version;
+	uint32_t latest_ver;
+	char devpath[BINARY_PATH_LEN];
+	user_binary_header_t header_data[PARTS_PER_BIN];
+
+	if (bin_idx < 0 || bin_idx > binary_manager_get_ucount()) {
+		bmdbg("Invalid bin idx %d\n", bin_idx);
+		return ERROR;
+	}
+
+	latest_ver = 0;
+	latest_idx = -1;
+	bin_count = 0;
+
+	for (part_idx = 0; part_idx < PARTS_PER_BIN; part_idx++) {
+		snprintf(devpath, BINARY_PATH_LEN, BINMGR_DEVNAME_FMT, BIN_PARTNUM(bin_idx, part_idx));
+		bmdbg("Devpath %s\n", devpath);
+		ret = binary_manager_read_user_header(devpath, &header_data[part_idx], crc_check);
+		if (ret == OK) {
+			bin_count++;
+			version = header_data[part_idx].bin_ver;
+			BIN_VER(bin_idx, part_idx) = version;
+			BIN_LOAD_PRIORITY(bin_idx, part_idx) = header_data[part_idx].loading_priority;
+			bmvdbg("Found valid header in part %d, version %d\n", part_idx, version);
+			if (version > latest_ver) {
+				latest_ver = version;
+				latest_idx = part_idx;
+			}
+		}
+	}
+
+	if (bin_count == 0) {
+		bmdbg("Failed to find valid header of binary %s\n", BIN_NAME(bin_idx));
+		return ERROR;
+	}
+
+	BIN_COUNT(bin_idx) = bin_count;
+	BIN_USEIDX(bin_idx) = latest_idx;
+	bmvdbg("Latest version %d, idx %d\n", latest_ver, latest_idx);
+
+	return OK;
 }
 
+int binary_manager_scan_common_bin(bool crc_check)
```

Comment:
This is also almost same as `binary_manager_scan_ubin`.
They can also be merged.

Reply:
Right, I merged them. Please refer the `binary_manager_scan_ubin_all`